app/uk/gov/hmrc/customs/notification/services/ClientWorkerImpl.scala

```diff
@@ -0,0 +1,172 @@
+                        callbackDetailsConnector: ApiSubscriptionFieldsConnector,
+                        pushConnector: PublicNotificationServiceConnector,
+                        pullConnector: NotificationQueueConnector,
+                        lockRepo: LockRepo,
+                        logger: NotificationLogger
+                      ) extends ClientWorker {
+
+  private val extendLockDuration =  org.joda.time.Duration.millis(config.pushNotificationConfig.lockRefreshDurationInMilliseconds)
+  private val refreshDuration = Duration(config.pushNotificationConfig.lockRefreshDurationInMilliseconds, TimeUnit.MILLISECONDS)
+
+  //TODO: should we pass in HeaderCarrier as an implicit parameter for logging?
+  override def processNotificationsFor(csid: ClientSubscriptionId, lockOwnerId: LockOwnerId): Future[Unit] = {
+    //implicit HeaderCarrier required for ApiSubscriptionFieldsConnector
+    //however looking at api-subscription-fields service I do not think it is required so keep new HeaderCarrier() for now
+    implicit val hc = HeaderCarrier()
+
+    val timer = actorSystem.scheduler.schedule(refreshDuration, refreshDuration, new Runnable {
+
+      override def run() = {
+        refreshLock(csid, lockOwnerId)
+      }
+    })
+
+    // cleanup timer
+    val eventuallyProcess = process(csid)
+    eventuallyProcess.onComplete { _ => // always cancel timer ie for both Success and Failure cases
+      logger.debug(s"about to cancel timer")
+      val cancelled = timer.cancel()
+      logger.debug(s"timer cancelled=$cancelled, timer.isCancelled=${timer.isCancelled}")
+    }
+
+    eventuallyProcess
+  }
+
+  private def refreshLock(csid: ClientSubscriptionId, lockOwnerId: LockOwnerId)(implicit hc: HeaderCarrier): Future[Unit] = {
+    lockRepo.tryToAcquireOrRenewLock(csid, lockOwnerId, extendLockDuration).map{ refreshedOk =>
+      if (!refreshedOk) {
+        val ex = new IllegalStateException("Unable to refresh lock")
+        throw ex
+      }
+    }.recover{
+      // If refresh of the lock fails there is nothing much we can do apart from logging the error
+      // It is unsafe to abort the notification processing as this could lead to the notifications
+      // database being in an inconsistent state eg notification could have been sent OK, but if
+      // we abort processing before notification is deleted then client could receive duplicate
+      // notifications
+      case e: Exception =>
+        val msg = e.getMessage
+        logger.error(msg) //TODO: extend logging API so that we can log an error on a throwable
+    }
+  }
+
+
+  protected def process(csid: ClientSubscriptionId)(implicit hc: HeaderCarrier): Future[Unit] = {
+
+    logger.info(s"About to process notifications")
+
+    (for {
+      clientNotifications <- repo.fetch(csid)
+      _ <- sequence(clientNotifications)(pushClientNotification)
+    } yield ())
+    .recover {
+      case e: Throwable =>
+        logger.error("Error pushing notification")
```

Comment:
here the error could be different then pushing, if we are failed to fetch the notifications then we should just exit from here and not send it to Pull.